### PR TITLE
Merge Feature/dev to development

### LINKED ISSUE #8: Refactor Makefile

### LINKED ISSUE #10: Integrate BandProtocol Oracle
This will be subdivided into multiple tasks

### PR DESCRIPTION
**Pull request by dheerajkd30**
Update include introduction of below new modules and sdk upgrades : 

1. cosmos-sdk v45.4
2. tendermint update to v0.34.19
3. Fee Grant Module: Allows paying fees on behalf of another account.
4. IBC v3 upgrade: This release includes v1 of the Interchain Accounts application module.
5. Authz module
6. cosmwasm update to v1.0.0 and permissioned cosmwasm update
7. Asset Module: The x/asset module includes asset creation and asset pair creation.
8. Vault Module: The vault module manages and stores the creation of vaults.
9. Locker Module: The Locker module allows a token holder to earn variable interest on their principal by locking their tokens into the Locker vault within the protocol. 
10. TokenMint Module: This module is used to mint tokens on a platform.
11. Rewards Module: Enables rewards on Liquidity Pools on AMMs and comodo.
12. Collector Module: Collector module is where all the fees and income generated from the protocol will get accumulated.
13. Liquidation Module: This triggers the liquidation of an asset if the Collateralization Ratio goes below the Liquidation ratio.
14. Band Oracle Module: The band oracle module fetches the prices of assets.
15. Market Module: A Market acts as a bridge between Real World Data and Blockchains.
16. Auction Module: This allows the platform to auction off the liquidated assets.
17. Liquidity Module: This module enables users on Comdex to Swap, transfer, and pool assets.
18. ESM Module : This module handles the logic of emergency shutdown mechanism.
19. Lend Module : This module handles the logic of borrowing and lending on comdex platform.
20. ICA : interchain accounts.
21. refactor makefile

### CODE REVIEW

File: proto/comdex/asset/v1beta1/asset.proto [hunk 1]
```diff
@@ -12,4 +12,5 @@ message Asset {
   string name    = 2 [ (gogoproto.moretags) = "yaml:\"name\"" ];
   string denom   = 3 [ (gogoproto.moretags) = "yaml:\"denom\"" ];
   int64 decimals = 4 [ (gogoproto.moretags) = "yaml:\"decimals\"" ];
+  bool isOnchain = 5 [ (gogoproto.moretags) = "yaml:\"isOnchain\"" ];
```
Comment (dheerajkd30) [inline proto/comdex/asset/v1beta1/asset.proto]: should be changed to is_on_chain in yaml. @pratikasr

File: x/asset/client/cli/tx.go [hunk 2]
```diff
@@ -160,27 +512,618 @@ func txUpdatePair() *cobra.Command {
...
+	}
+	genesisSupply, err := ParseStringFromString(assetMapping.GenesisSupply, ",")
+	if err != nil {
+		return txf, nil, err
+	}
+	isGovToken, err := ParseStringFromString(assetMapping.IsGovToken, ",")
+	if err != nil {
+		return txf, nil, err
+	}
+	recipient, err := ParseStringFromString(assetMapping.Recipient, ",")
```
Comment (dheerajkd30) [inline x/asset/client/cli/tx.go:490]: are we expecting multiple recipients ? if not can we not use string instead @cgsingh33 @rajrohit10

File: x/asset/client/cli/tx.go [hunk 3]
```diff
@@ -160,27 +512,618 @@ func txUpdatePair() *cobra.Command {
...
+
+	assetId, err := ParseUint64SliceFromString(assetMapping.AssetId, ",")
+	if err != nil {
+		return txf, nil, err
+	}
+	genesisSupply, err := ParseStringFromString(assetMapping.GenesisSupply, ",")
+	if err != nil {
+		return txf, nil, err
+	}
+	isGovToken, err := ParseStringFromString(assetMapping.IsGovToken, ",")
```
Comment (dheerajkd30) [inline x/asset/client/cli/tx.go:486]: any reason we are not using string instead @cgsingh33 @rajrohit10

File: types/utils.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+// AddOrSet Set when the key not existed on the map or add existed value of the key.
+func (m StrIntMap) AddOrSet(key string, value sdk.Int) {
+	if _, ok := m[key]; !ok {
+		m[key] = value
+	} else {
+		m[key] = m[key].Add(value)
+	}
+}
+
+func PP(data interface{}) {
```
Comment (dheerajkd30) [inline types/utils.go:34]: unused code to be removed. @cgsingh33

Comment (dheerajkd30) [inline types/utils.go:34]: please remove if unused. @cgsingh33

File: x/asset/client/cli/flags.go [hunk 5]
```diff
@@ -3,13 +3,27 @@ package cli
...
 const (
-	flagLiquidationRatio       = "liquidation-ratio"
-	flagName                   = "name"
-	flagDenom                  = "denom"
-	flagDecimals               = "decimals"
+	flagLiquidationRatio        = "liquidation-ratio"
+	flagName                    = "name"
+	flagDenom                   = "denom"
+	flagDecimals                = "decimals"
+	flagCollateralWeight        = "collateralWeight"
```
Comment (dheerajkd30) [inline x/asset/client/cli/flags.go]: collateral-weight
liquidation-threshold
is-bridged-asset , lets follow similar naming style across. @cgsingh33 @pratikasr @rajrohit10

File: x/asset/client/cli/flags.go [hunk 6]
```diff
@@ -19,4 +33,118 @@ func GetLiquidationRatio(cmd *cobra.Command) (sdk.Dec, error) {
 	}
 
 	return sdk.NewDecFromStr(s)
-}
\ No newline at end of file
+}
+
+func ParseStringFromString(s string, seperator string) ([]string, error) {
```
Comment (dheerajkd30) [inline x/asset/client/cli/flags.go:38]: confusing, shouldn't this be ParseStringFromStringArray/slice

File: x/asset/client/cli/query.go [hunk 7]
```diff
@@ -187,3 +187,291 @@ func queryParams() *cobra.Command {
...
+
+	flags.AddQueryFlagsToCmd(cmd)
+	flags.AddPaginationFlagsToCmd(cmd, "assets")
+
+	return cmd
+}
+
+func queryTokenGov() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "check_for_token_gov [app_id]",
```
Comment (dheerajkd30) [inline x/asset/client/cli/query.go]: check-for-token-gov @cgsingh33

Comment (dheerajkd30) [reply x/asset/client/cli/query.go]: or better app-gov-token

File: x/asset/client/cli/tx.go [hunk 8]
```diff
@@ -1,58 +1,114 @@
...
+			if err = msg.ValidateBasic(); err != nil {
+				return err
+			}
+
+			return tx.GenerateOrBroadcastTxCLI(clientCtx, cmd.Flags(), msg)
 		},
 	}
 
-	flags.AddTxFlagsToCmd(cmd)
+	cmd.Flags().String(cli.FlagTitle, "", "title of proposal")
```
Comment (dheerajkd30) [inline x/asset/client/cli/tx.go:102]: we have been using it a lot , lets move strings to constants.

File: x/asset/client/cli/tx.go [hunk 9]
```diff
@@ -77,81 +133,377 @@ func txUpdateAsset() *cobra.Command {
...
```
Comment (dheerajkd30) [inline x/asset/client/cli/tx.go:185]: need to refactor to constants everywhere.

File: x/asset/client/cli/tx.go [hunk 10]
```diff
@@ -77,81 +133,377 @@ func txUpdateAsset() *cobra.Command {
...
+	cmd.Flags().String(cli.FlagDeposit, "", "deposit of proposal")
+	_ = cmd.MarkFlagRequired(cli.FlagTitle)
+	_ = cmd.MarkFlagRequired(cli.FlagDescription)
+
+	return cmd
+}
+
+func NewCmdSubmitAddWhitelistedAssetsProposal() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "add-lend-assets [asset_id] [Collateral_Weight] [Liquidation_Threshold] [Is_Bridged_Asset]",
```
Comment (dheerajkd30) [inline x/asset/client/cli/tx.go]: [refactor ](https://github.com/comdex-official/comdex/issues/236). @dheerajkd30

File: x/asset/client/cli/tx.go [hunk 11]
```diff
@@ -77,81 +133,377 @@ func txUpdateAsset() *cobra.Command {
...
+			description, err := cmd.Flags().GetString(cli.FlagDescription)
+			if err != nil {
+				return err
+			}
+
+			from := ctx.GetFromAddress()
+
+			var assets []types.ExtendedAsset
+			for i := range assetId {
+				newcollateralWeigt, err := sdk.NewDecFromStr(collateralWeight[i])
```
Comment (dheerajkd30) [inline x/asset/client/cli/tx.go]: spell check newcollateralWeigt @cgsingh33

File: x/asset/client/cli/tx.go [hunk 12]
```diff
@@ -77,81 +133,377 @@ func txUpdateAsset() *cobra.Command {
...
+				if err != nil {
+					return err
+				}
+				newliquidationThreshold, err := sdk.NewDecFromStr(liquidationThreshold[i])
+				if err != nil {
+					return err
+				}
+				newisBridgedAsset := ParseBoolFromString(isBridgedAsset[i])
```
Comment (dheerajkd30) [inline x/asset/client/cli/tx.go]: refactor @cgsingh33

File: x/asset/client/cli/tx.go [hunk 13]
```diff
@@ -77,81 +133,377 @@ func txUpdateAsset() *cobra.Command {
...
+			newcollateralWeight, err := sdk.NewDecFromStr(collateralWeight)
+			if err != nil {
+				return err
+			}
+
+			liquidationThreshold, err := cmd.Flags().GetString(flagLiquidationThreshold)
+			if err != nil {
+				return err
+			}
+			newliquidationThreshold, err := sdk.NewDecFromStr(liquidationThreshold)
```
Comment (dheerajkd30) [inline x/asset/client/cli/tx.go]: refractor

File: x/asset/client/cli/tx.go [hunk 14]
```diff
@@ -77,81 +133,377 @@ func txUpdateAsset() *cobra.Command {
...
+			id, err := strconv.ParseUint(args[0], 10, 64)
+			if err != nil {
+				return err
+			}
+
+			collateralWeight, err := cmd.Flags().GetString(flagCollateralWeight)
+			if err != nil {
+				return err
+			}
+			newcollateralWeight, err := sdk.NewDecFromStr(collateralWeight)
```
Comment (dheerajkd30) [inline x/asset/client/cli/tx.go]: refractor

File: x/asset/client/cli/tx.go (hunk 2 above)
Comment (cgsingh33) [reply x/asset/client/cli/tx.go:490]: there will be only one recipient for an app and asset. But using this proposal we can add multiple asset mapping for an app, that is why we used comma-separated values.

File: x/asset/client/cli/tx.go (hunk 3 above)
Comment (cgsingh33) [reply x/asset/client/cli/tx.go:486]: same as above for adding proposal for multiple asset mapping.

**Review by bsrinivas8687 — DISMISSED** [review]

**Review by bsrinivas8687 — APPROVED** [review]

**Pull request merged**